src/transports/dash/add_segment_integrity_checks_to_loader.ts

```diff
@@ -47,14 +59,56 @@ export default function addSegmentIntegrityChecks(
     ITransportTextSegmentLoader |
     ITransportImageSegmentLoader
 {
-  return (content) => segmentLoader(content).pipe(tap((res) => {
-    if ((res.type === "data-loaded" || res.type === "data-chunk") &&
-        res.value.responseData !== null &&
-        typeof res.value.responseData !== "string" &&
-        !isWEBMEmbeddedTrack(content.representation))
+  return (content) => segmentLoader(content).pipe(mergeMap((res) => {
+    if (res.type !== "data-loaded" && res.type !== "data-chunk") {
+      return observableOf(res);
+    }
+
+    if (res.value.responseData === null ||
+        typeof res.value.responseData === "string" ||
+        isWEBMEmbeddedTrack(content.representation))
     {
-      checkISOBMFFIntegrity(new Uint8Array(res.value.responseData),
-                            content.segment.isInit);
+      return observableOf(res);
     }
+
+    let responseDataBytes = new Uint8Array(res.value.responseData);
+
+    // Check if we might have done a byte-range request
+    if (content.segment.range !== undefined &&
+        (content.segment.range[1] - content.segment.range[0] + 1)
+          === responseDataBytes.length)
+    {
+      const numberOfMissingBytes = getMissingBytes(responseDataBytes,
+                                                   content.segment.isInit);
+
+      if (numberOfMissingBytes !== undefined && numberOfMissingBytes !== 0) {
+        if (numberOfMissingBytes < 0) {
+          responseDataBytes = responseDataBytes
+            .slice(0, responseDataBytes.length + numberOfMissingBytes);
+        }
+
+        // there's a positive amount of bytes missing.
+        // Redo the request directly with a corrected `segment` object
+        const oldRange = content.segment.range;
+        const newRange = [oldRange[0], oldRange[1] + numberOfMissingBytes];
+        const newSegment = objectAssign({},
+                                        content.segment,
+                                        { range: newRange });
+        const newContent = objectAssign({},
+                                        content,
+                                        { segment: newSegment });
+        const newRequest$ = segmentLoader(newContent).pipe(catchError((err) => {
+          log.warn("DASH: retried segment request with another byte range failed",
```

Comment:
Would it be better to filter on errors before return the origin response ?

Reply:
I'm not sure to understand what you mean.

Reply:
Maybe this could fail because of a network error, or something like this. Maybe we would try to re-do the request before returning the original response.